src/qz/printer/info/NativePrinter.java

```diff
@@ -0,0 +1,180 @@
+    private PrinterProperty<PrinterResolution> resolution;
+    private PrinterProperty<String> driverFile;
+
+    public NativePrinter(String printerId) {
+        this.printerId = printerId;
+        this.description = new PrinterProperty<>();
+        this.printService = new PrinterProperty<>();
+        this.driverFile = new PrinterProperty<>();
+        this.driver = new PrinterProperty<>();
+        this.resolution = new PrinterProperty<>();
+        this.outdated = false;
+    }
+
+    public PrinterProperty<String> getDescription() {
+        return description;
+    }
+
+    public void setDescription(String description) {
+        this.description.set(description);
+    }
+
+    public PrinterProperty<String> getDriverFile() {
+        return driverFile;
+    }
+
+    public void setDriverFile(String driverFile) {
+        this.driverFile.set(driverFile);
+    }
+
+    public PrinterProperty<String> getDriver() {
+        if (!driver.isSet()) {
+            getDriverAttributes(this);
+        }
+        return driver;
+    }
+
+    public String getName() {
+        if (printService != null && printService.value() != null) {
+            return printService.value().getName();
+        }
+        return null;
+    }
+
+    public PrinterName getLegacyName() {
+        if (printService != null && printService.value() != null) {
+            return printService.value().getAttribute(PrinterName.class);
+        }
+        return null;
+    }
+
+    public void setDriver(String driver) {
+        this.driver.set(driver);
+    }
+
+    public PrinterProperty<PrintService> getPrintService() {
+        return printService;
+    }
+
+    public void setPrintService(PrintService printService) {
+        // Fetch resolution, if available
+        Object resolution = printService.getDefaultAttributeValue(PrinterResolution.class);
+        if (resolution != null) {
+            this.resolution.set((PrinterResolution)resolution);
+        }
+        this.printService.set(printService);
+    }
+
+    public String getPrinterId() {
+        return printerId;
+    }
+
+    public void setResolution(PrinterResolution resolution) {
+        this.resolution.set(resolution);
+    }
+
+    public PrinterProperty<PrinterResolution> getResolution() {
+        if (!resolution.isSet()) {
+            getDriverAttributes(this);
+        }
+        return resolution;
+    }
+
+    public List<Integer> getResolutions() {
+        // TODO: Test/Implement supported resolutions for CUPS
+        List<Integer> densities = new ArrayList<>();
+
+        PrintService ps = getPrintService().value();
+        PrinterResolution[] resSupport = (PrinterResolution[])ps.getSupportedAttributeValues(PrinterResolution.class, ps.getSupportedDocFlavors()[0], null);
+        if (resSupport == null || resSupport.length == 0) {
+            resSupport = new PrinterResolution[]{ getResolution().value() };
+            }
+        if (resSupport != null) {
```

Comment:
Won't `resSupport` never be null at this point due to the preceding if statement?

Reply:
Done.